### PR TITLE
feat: pwa registration

### LINKED ISSUE #637: SW activation and cache static files
## Description

We have some static files those safe to be cache on the client side.

## Implementation Model

We can use [next-pwa](https://github.com/shadowwalker/next-pwa)

To be cache:

- All hashed CSS and JS - StaleWhileRevalidate/CacheFirst
- All Cloudinary image - CacheFirst with 30d expiry

DO NOT cache any HTML files

## Tasks

- [ ] Install `next-pwa`
- [ ] Add the next-pwa configuration on seperate file: `workbox.config.js`
- [ ] Link to the `next.config.js`
- [ ] Custom Install Button -- optional

### PR DESCRIPTION
**Pull request by svspicious**
Closes #637 

Related to #716 #719 

## Description

I am new with CSP stuff, but I think this might be a clue https://qubyte.codes/blog/content-security-policy-and-service-workers.

TLDR; SW is not using `image-src` policy but using `connect-src` policy instead. Since we don't have `connect-src` it fallback to `default-src` wich is only limited to `self` / origin of the domain itself. So, it refuses to **_"connect"_**

image

Mas @zainfathoni tried to add Cloudinary url to `script-src` (at #719 ) but ofc it doesn't work, we need to add to `connect-src`.

So I tried to add res.cloudinary.com to `connect-src` and deploy to my personal netlify account for preview (yes, I'm dumb. brb making a Draft PR for deploy preview). And it works (and it caches cloudinary img assets)

https://hardcore-noyce-3b9626.netlify.app/

image
image

## Current Tasks

- [x] implement @mazipan PWA registration (#705 )
- [x] Add CSP rules for cloudinary assets

### CODE REVIEW

**Comment by svspicious** [comment]
PWA lighthouse report
image

File: netlify.toml [hunk 1]
```diff
@@ -12,7 +12,7 @@
     X-XSS-Protection = "1; mode=block"
     X-Content-Type-Options = "nosniff"
     Referrer-Policy = "same-origin"
-    Content-Security-Policy = "default-src 'self'; connect-src 'self' analytics.google.com stats.g.doubleclick.net; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' *.googletagmanager.com analytics.google.com; img-src 'self' blob: data: https:; frame-ancestors 'none';"
+    Content-Security-Policy = "default-src 'self'; connect-src 'self' analytics.google.com stats.g.doubleclick.net res.cloudinary.com; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' *.googletagmanager.com analytics.google.com; img-src 'self' blob: data: https:; frame-ancestors 'none';"
     Permissions-Policy = "camera=(), microphone=(), geolocation=(), interest-cohort=()"
 
 [[headers]]
```
Comment (mazipan) [inline netlify.toml:18]: I just remember, we need to make sure `/service-worker.js` never send any cache header.
I think we need to add new header configuration and set it  as `no-store, max-age=0`

**Comment by mazipan** [comment]
Need help from @zainfathoni and others also to test the preview domain.
I believe the error can not be catch by Lighthouse test.
We need to check manually.

**Comment by ekafyi** [comment]
I just checked https://deploy-preview-762--wargabantuwarga.netlify.app/ and got mixed result.

Most images were successfully fetched and cached by the SW as intended.

Some other images (hero_banner and sembako-cta) were not.

Screen Shot 2021-08-18 at 22 41 08

I initially _thought_ it had something to do with images rendered via Next's `Image` component, but the telemedicine banner, which also uses `Image` component, was fetched correctly.

🤔

**Review by mazipan — APPROVED** [review]

File: netlify.toml [hunk 2]
```diff
@@ -25,6 +25,11 @@
     Content-Security-Policy = "default-src 'self'; connect-src 'self' analytics.google.com stats.g.doubleclick.net; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' *.googletagmanager.com *.netlify.com unpkg.com analytics.google.com 'unsafe-eval' media-library.cloudinary.com; frame-src 'self' cloudinary.com; img-src 'self' blob: data: https:; frame-ancestors 'none';"
     Permissions-Policy = "camera=(), microphone=(), geolocation=(), interest-cohort=()"
 
+[[headers]]
+  for = "/service-worker.js"
+  [headers.values]
+    Cache-Control = "no-store, max-age=0"
```
Comment (resir014) [inline netlify.toml:31]: Don't forget to duplicate this header into `next.config.js` as well, just in case `@netlify/plugin-nextjs` supports headers/rewrites directly from `next.config.js` in the near future.

**Review by resir014 — CHANGES_REQUESTED** [review]

File: netlify.toml (hunk 2 above)
Comment (svspicious) [reply netlify.toml:31]: got it

**Review by resir014 — APPROVED** [review]
🚢

**Review by zainfathoni — APPROVED** [review]
LGTM 💯 

Thanks for working on it, Mas @svspicious! 🙏
